### PR TITLE
v10 matrix friendly API additions

### PR DESCRIPTION
**Pull request by simonbowly**
Things to add/update for the matrix API classes:

- [x] MVar: fromlist, fromvar, transpose, T, reshape, item, diagonal
- [x] MLinExpr: sum, clear, ndim, size, item, zeros
- [x] MQuadExpr: sum, clear, ndim, size, item, zeros
- [x] New MQConstr class: needs a new overload MQuadExpr -> TempMQConstr (fake) -> MQConstr
- [x] deprecated MVar constructor (delete from stubs?)
- [x] New operator overloads which weren't previously handled

### CODE REVIEW

File: gurobipy-stubs/gurobipy.pyi [hunk 1]
```diff
@@ -815,7 +815,29 @@ class MConstr:
...
+    @property
+    def ndim(self) -> int: ...
+    @property
+    def shape(self) -> Tuple[int, ...]: ...
+    @property
+    def size(self) -> int: ...
+
 class MLinExpr:
+    @classmethod
+    def zeros(self, shape: Union[int, Tuple[int, ...]]) -> MLinExpr: ...
```
Comment (rluce) [inline gurobipy-stubs/gurobipy.pyi]: The `Union[int, Tuple[int, ...]])` appears often enough that it warrents its own type alias, maybe `ShapeLike`.  Please check the NumPy type hints, I'm certain that they have something for exactly this case, and we could align with that.

**Review by rluce — CHANGES_REQUESTED** [review]
I think we need a more overloads in the arithmetic methods for term-based modeling operands, Var, LinExpr, QuadExpr

File: gurobipy-stubs/gurobipy.pyi (hunk 1 above)
Comment (simonbowly) [reply gurobipy-stubs/gurobipy.pyi]: No ShapeLike equivalent listed [in numpy docs](https://numpy.org/devdocs/reference/typing.html#), so I wrote an alias `_ShapeLike` into our stubs

**Comment by simonbowly** [comment]
@rluce I added the scalar equivalents via new aliases `_ConstComponent`, `_LinearComponent`, `QuadComponent`. This seemed the simplest way to handle the MVar -> MLinExpr -> MQuadExpr promotion based on expression degree.

**Review by rluce — APPROVED** [review]

**Comment by rluce** [comment]
Great, thanks!

**Pull request merged**